src/flashcart/ed64/ed64_ll.c

```diff
@@ -0,0 +1,409 @@
+void pi_dma_from_cart (void* dest, void* src, unsigned long size);
+void pi_dma_to_cart (void* dest, void* src, unsigned long size);
+void pi_dma_from_sram (void *dest, unsigned long offset, unsigned long size);
+void pi_dma_to_sram (void* src, unsigned long offset, unsigned long size);
+void pi_dma_from_cart_safe (void *dest, void *src, unsigned long size);
+
+void ed64_ll_set_sdcard_timing (void);
+
+
+#define SAV_EEP_ON 1
+#define SAV_SRM_ON 2
+#define SAV_EEP_SIZE 4
+#define SAV_SRM_SIZE 8
+
+#define SAV_RAM_BANK 128
+#define SAV_RAM_BANK_APPLY 32768
+
+uint32_t ed64_ll_reg_read (uint32_t reg);
+void ed64_ll_reg_write (uint32_t reg, uint32_t data);
+
+uint8_t ed64_ll_sram_bank;
+ed64_save_type_t ed64_ll_save_type;
+
+
+uint32_t ed64_ll_reg_read (uint32_t reg) {
+
+    *(volatile uint32_t *) (ED64_CONFIG_REGS_BASE);
+    return *(volatile uint32_t *) (ED64_CONFIG_REGS_BASE + reg * 4);
+
+}
+
+void ed64_ll_reg_write (uint32_t reg, uint32_t data) {
+
+    *(volatile uint32_t *) (ED64_CONFIG_REGS_BASE);
+    *(volatile uint32_t *) (ED64_CONFIG_REGS_BASE + reg * 4) = data;
+    *(volatile uint32_t *) (ROM_ADDRESS);
+
+}
+
+
+ed64_save_type_t ed64_ll_get_save_type (void) {
+
+    return ed64_ll_save_type;
+
+}
+
+void ed64_ll_set_save_type (ed64_save_type_t type) {
+
+    uint16_t save_cfg;
+    uint8_t eeprom_on, sram_on, eeprom_size, sram_size, ram_bank;
+    ed64_ll_save_type = type;
+    eeprom_on = 0;
+    sram_on = 0;
+    eeprom_size = 0;
+    sram_size = 0;
+    ram_bank = ed64_ll_sram_bank;
+
+
+    switch (type) {
+        case SAVE_TYPE_EEPROM_16K:
+            eeprom_on = 1;
+            eeprom_size = 1;
+            break;
+        case SAVE_TYPE_EEPROM_4K:
+            eeprom_on = 1;
+            break;
+        case SAVE_TYPE_SRAM:
+            sram_on = 1;
+            break;
+        case SAVE_TYPE_SRAM_128K:
+            sram_on = 1;
+            sram_size = 1;
+            break;
+        case SAVE_TYPE_FLASHRAM:
+            sram_on = 0;
+            sram_size = 1;
+            break;
+        default:
+            sram_on = 0;
+            sram_size = 0;
+            ram_bank = 1;
+            break;
+    }
+
+    if (eeprom_on)save_cfg |= SAV_EEP_ON;
+    if (sram_on)save_cfg |= SAV_SRM_ON;
+    if (eeprom_size)save_cfg |= SAV_EEP_SIZE;
+    if (sram_size)save_cfg |= SAV_SRM_SIZE;
+    if (ram_bank)save_cfg |= SAV_RAM_BANK;
+    save_cfg |= SAV_RAM_BANK_APPLY;
+
+    ed64_ll_reg_write(REG_SAV_CFG, save_cfg);
+
+}
+
+void ed64_ll_set_sram_bank (uint8_t bank) {
+
+    ed64_ll_sram_bank = bank == 0 ? 0 : 1;
+
+}
+
+
+void pi_initialize (void) {
+
+	dma_wait();
+	io_write(PI_STATUS_REG, 0x03);
+
+}
+
+// Inits PI for sram transfer
+void pi_initialize_sram (void) {
+
+	io_write(PI_BSD_DOM2_LAT_REG, 0x05);
+	io_write(PI_BSD_DOM2_PWD_REG, 0x0C);
+	io_write(PI_BSD_DOM2_PGS_REG, 0x0D);
+	io_write(PI_BSD_DOM2_RLS_REG, 0x02);
+
+}
+
+void pi_dma_from_sram (void *dest, unsigned long offset, unsigned long size) {
```

Comment:
please use `pi_dma_read_data`/`pi_dma_write_data` functions from `flashcart_utils.c` instead of these functions that do not use libdragon provided implementations

Reply:
the reason i didnt use the libdragon implimentation is because the sram address is write protected (i think). it results in this 
![image](https://github.com/Polprzewodnikowy/N64FlashcartMenu/assets/45049787/a7632fec-a610-4cda-919e-0ba1aec9a2db)

Reply:
After adding `#include "../flashcart_utils.h"`
Try using `pi_dma_read_data((void*)(ED64_SAVE_ADDR_BASE + address_offset), buffer, size);`